### PR TITLE
issue #5: Remanufacture the code to use a single dataframe

### PR DESCRIPTION
**Pull request by aquintanas**

### CODE REVIEW

File: p2p.py [hunk 1]
```diff
@@ -63,11 +65,36 @@ def read_marketplace_files(data_directory, marketplace_name):
 def collect_investment_data(data_directory):
     data_files = {}
     investment_platforms = [entry.name for entry in os.scandir(data_directory) if entry.is_dir()]
+    print(investment_platforms)
```
Comment (limdor) [inline p2p.py:68]:
```suggestion
```

**Review by limdor — COMMENTED** [review]
Nice improving of the report but we should keep some information that with this change we are loosing.
In the overall investments before we have:
```
*****************************
**** Overall Investments ****
*****************************
The portfolio consists of at least 100 different loan parts: TODO
Overall Investment on 2021-08-31: 27190.08€
                Outstanding principal  Percentage
Country                                          
Bulgaria                 12275.943538    0.451486
Latvia                    3117.150000    0.114643
Romania                   2788.613814    0.102560
Spain                     2484.710000    0.091383
Poland                    2396.725448    0.088147
Moldova                   1460.570000    0.053717
United Kingdom             956.430000    0.035176
Estonia                    925.730000    0.034047
Lithuania                  784.210000    0.028842
No more than 50% of loans are issued in 3 (or less) countries: 66.87%
No more than 33% of loans are issued in any single country: 45.15%
Overall Investment on 2021-08-31: 27190.08€
                   Outstanding principal  Percentage
Loan originator                                     
AccessFinance                8489.446333    0.312226
EasyCredit                   3736.367205    0.137417
iCredit                      3522.909262    0.129566
ID Finance                   2484.710000    0.091383
Eleving Group                2101.550000    0.077291
Sun Finance                  1618.970000    0.059543
DelfinGroup                  1498.180000    0.055100
Mozipo Group                  950.400000    0.034954
Novaloans                     950.020000    0.034940
ESTO                          925.730000    0.034047
Mikro Kapital                 805.260000    0.029616
CashCredit                     50.130000    0.001844
FastFinance                    50.000000    0.001839
Evergreen Finance               6.410000    0.000236
No more than 50% of loans are issued by 5 (or less) lending companies: 74.79%
No more than 20% of loans are issued by any single lending company: 31.22%
```
I think this information is valuable and we should keep it, and we can add a new section for the report. We could leave it like:
```
*****************************
**** Overall Investments ****
*****************************
Overall Investment on 2021-08-31: 27190.08€
                Outstanding principal  Percentage
Country                                          
Bulgaria                 12275.943538    0.451486
Latvia                    3117.150000    0.114643
Romania                   2788.613814    0.102560
Spain                     2484.710000    0.091383
Poland                    2396.725448    0.088147
Moldova                   1460.570000    0.053717
United Kingdom             956.430000    0.035176
Estonia                    925.730000    0.034047
Lithuania                  784.210000    0.028842
Overall Investment on 2021-08-31: 27190.08€
                   Outstanding principal  Percentage
Loan originator                                     
AccessFinance                8489.446333    0.312226
EasyCredit                   3736.367205    0.137417
iCredit                      3522.909262    0.129566
ID Finance                   2484.710000    0.091383
Eleving Group                2101.550000    0.077291
Sun Finance                  1618.970000    0.059543
DelfinGroup                  1498.180000    0.055100
Mozipo Group                  950.400000    0.034954
Novaloans                     950.020000    0.034940
ESTO                          925.730000    0.034047
Mikro Kapital                 805.260000    0.029616
CashCredit                     50.130000    0.001844
FastFinance                    50.000000    0.001839
Evergreen Finance               6.410000    0.000236
*********************************
**** Diversification reports ****
*********************************
Report on 2021-08-31
|- Overall Investment: 27190.08€
|- The portfolio consists of at least 100 different loan parts: 669
|- Statistics by Country:
   |- No more than 50% of loans are issued in 3 (or less) countries: 66.87%
   |- No more than 33% of loans are issued in any single country: 45.15%
|- Statistics by Originator:
   |- No more than 50% of loans are issued by 5 (or less) lending companies: 70.65%
   |- No more than 20% of loans are issued by any single lending company: 31.22%
```
Last but not least, it should consider if the -p parameter is passed when the program is called. Currently the idea was that it shows only data of the last month if it is not passed and all the history if it is passed.

**Comment by aquintanas** [comment]
> Last but not least, it should consider if the -p parameter is passed when the program is called. Currently the idea was that it shows only data of the last month if it is not passed and all the history if it is passed.

A prior, I think in two options: (a) check if the argument exists in situ, i.e. every time that we want to compute some statistic, or (b) filter the files at the very beginning according to the kind of report (all or latest). 

Despite the method, I doubt that when you select only to show the latest data, do you refer to the last file or the last/newest date? I mean, if the last file of Mintos is from X and Iuvos is from Y, the latest/newest data report should only report Mintos or both platforms? 

If only Mintos has to be shown, I think that filter the files at the very beginning (b) is a good option.

**Comment by aquintanas** [comment]
@limdor, can you review the code? I think that the code now fulfils your requirements.

**Comment by aquintanas** [comment]
@limdor ,  Now the result display follows your suggestions: 

````
*****************************
**** Overall Investments ****
*****************************
Overall Investments on 2021-08-31
|- Overall Investment: 27190.08€
|- Investment by Country
                Outstanding principal  Percentage
Country                                          
Bulgaria                 12275.943538    0.451486
Latvia                    3117.150000    0.114643
Romania                   2788.613814    0.102560
Spain                     2484.710000    0.091383
Poland                    2396.725448    0.088147
Moldova                   1460.570000    0.053717
United Kingdom             956.430000    0.035176
Estonia                    925.730000    0.034047
Lithuania                  784.210000    0.028842
|- Investment by Country
                   Outstanding principal  Percentage
Loan originator                                     
AccessFinance                8489.446333    0.312226
EasyCredit                   3736.367205    0.137417
iCredit                      3522.909262    0.129566
ID Finance                   2484.710000    0.091383
Eleving Group                2101.550000    0.077291
Sun Finance                  1618.970000    0.059543
DelfinGroup                  1498.180000    0.055100
Mozipo Group                  950.400000    0.034954
Novaloans                     950.020000    0.034940
ESTO                          925.730000    0.034047
Mikro Kapital                 805.260000    0.029616
CashCredit                     50.130000    0.001844
FastFinance                    50.000000    0.001839
Evergreen Finance               6.410000    0.000236
*********************************
**** Diversification reports ****
*********************************
Investments diversification on 2021-08-31
|- Diversification Investment: 27190.08€
|- The portfolio consists of at least 100 different loan parts: 669
|- Statistics by Country:
   |- No more than 50% of loans are issued in 3 (or less) countries: 66.87%
   |- No more than 33% of loans are issued in any single country: 45.15%
|- Statistics by Originator:
   |- No more than 50% of loans are issued by 5 (or less) lending companies: 74.79%
   |- No more than 20% of loans are issued by any single lending company: 31.22%
````

Can you review the code? If you find that it is ok, we can close this PR by merging it into the master branch.

**Review by limdor — APPROVED** [review]

File: p2p.py [hunk 2]
```diff
@@ -84,66 +125,110 @@ def report(show_past_investments):
...
+        print(f"|- Statistics by Originator:")
+        overall_group_by_originator = overall_data['DataByLoanOriginator']
+
+        sum_on_top_5_originators = overall_group_by_originator[OUTSTANDING_PRINCIPAL][0:5].sum()
+        percentage_top_5_originators = (sum_on_top_5_originators / total_invested_by_date) * 100
+        print(f'   |- No more than 50% of loans are issued by 5 (or less) lending companies: {percentage_top_5_originators:.2f}%')
+
+        top_originator = overall_group_by_originator[OUTSTANDING_PRINCIPAL][0]
+        percentage_top_originator = (top_originator / total_invested_by_date) * 100
+        print(f'   |- No more than 20% of loans are issued by any single lending company: {percentage_top_originator:.2f}%')
```
[the remark this answers is not in the document]
Comment (aquintanas) [reply p2p.py:234]: I will add it in the following issue or as a hotfix!

**Pull request merged**